pkg/auth/serviceaccount.go

```diff
@@ -101,6 +101,25 @@ func EnsureOIDCServiceAccountExistsForResource(ctx context.Context, serviceAccou
 	return nil
 }
 
+// DeleteOIDCServiceAccountIfExists makes sure the given resource does not have an OIDC service account.
+// If it does that service account is deleted.
+func DeleteOIDCServiceAccountIfExists(ctx context.Context, serviceAccountLister corev1listers.ServiceAccountLister, kubeclient kubernetes.Interface, gvk schema.GroupVersionKind, objectMeta metav1.ObjectMeta) error {
+	saName := GetOIDCServiceAccountNameForResource(gvk, objectMeta)
+	sa, err := serviceAccountLister.ServiceAccounts(objectMeta.Namespace).Get(saName)
+
+	// Service Account Found
+	if sa != nil {
+		logging.FromContext(ctx).Debugw("OIDC Service account already exists. Deleting OIDC service account")
+
+		err = kubeclient.CoreV1().ServiceAccounts(objectMeta.Namespace).Delete(ctx, sa.Name, metav1.DeleteOptions{})
```

Comment:
I'm not sure if this is needed, but mybe we can check the `OwnerReference` on the service account before deleting it? That way, on the off chance that the OIDC name we have chosen for the service account clashes with an existing service account we can know that we made the service account when we delete, and aren't deleting a user-created service account.

WDYT @creydr ?

Reply:
That's a good idea

Reply:
So for this we update the if condition of `DeleteOIDCServiceAccountIfExists` to
```
	if !apierrs.IsNotFound(err) && metav1.IsControlledBy(&sa.ObjectMeta, &objectMeta) {
```
so that we don't delete the user created serviceAccount and consider that there are no serviceAccounts, right?

Reply:
Hey @pawarpranav83,
yes. 
But I guess you should use
```
if err == nil && metav1.IsControlledBy(&sa.ObjectMeta, &objectMeta) {
```
in line 111. Because if there was another error ( != NotFound), then `sa` can be `nil` and this would cause a panic in your case.